File: app/src/main/java/com/github/keeganwitt/applist/ApplicationInfoUtils.java
package com.github.keeganwitt.applist;

import android.app.usage.StorageStats;
import android.app.usage.StorageStatsManager;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.icu.text.DateFormat;
import android.os.Build;
import android.os.Environment;
import android.os.storage.StorageManager;
import android.os.storage.StorageVolume;
import android.text.format.Formatter;
import android.util.Log;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import static android.os.Process.myUserHandle;

public class ApplicationInfoUtils {
    private static final String TAG = ApplicationInfoUtils.class.getSimpleName();

    private static Map<String, Long> lastUsedEpochsCache;

    private ApplicationInfoUtils() {}

    @SuppressWarnings("deprecation")
    public static String getPackageInstaller(PackageManager packageManager, ApplicationInfo applicationInfo) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            try {
                return packageManager.getInstallSourceInfo(applicationInfo.packageName).getInstallingPackageName();
            } catch (PackageManager.NameNotFoundException e) {
                Log.e(TAG, "Unable to determine package installer for " + applicationInfo.packageName, e);
                return null;
            }
        } else {
            return packageManager.getInstallerPackageName(applicationInfo.packageName);
        }
    }

    public static String getPackageInstallerName(String installerPackageName) {
        if ("com.amazon.venezia".equals(installerPackageName)) {
            return "Amazon Appstore";
        } else if ("cm.aptoide.pt".equals(installerPackageName)) {
            return "Aptoide";
        } else if ("net.rim.bb.appworld".equals(installerPackageName)) {
            return "Blackberry World";
        } else if ("com.farsitel.bazaar".equals(installerPackageName)) {
            return "Cafe Bazaar";
        } else if ("com.sec.android.app.samsungapps".equals(installerPackageName)) {
            return "Galaxy Store";
        } else if ("com.android.vending".equals(installerPackageName)) {
            return "Google Play";
        } else if ("com.huawei.appmarket".equals(installerPackageName)) {
            return "Huawei App Galary";
        } else if ("com.xiaomi.market".equals(installerPackageName)) {
            return "Mi Store";
        } else if ("com.oneplus.backuprestore".equals(installerPackageName)) {
            return "OnePlus Clone Phone";
        } else if ("com.sec.android.easyMover".equals(installerPackageName)) {
            return "Samsung Smart Switch";
        } else if ("com.slideme.sam.manager".equals(installerPackageName)) {
            return "SlideME Marketplace";
        } else if ("com.tencent.android.qqdownloader".equals(installerPackageName)) {
            return "TenCent Appstore";
        } else if ("com.yandex.store".equals(installerPackageName)) {
            return "Yandex Appstore";
        } else if (installerPackageName != null) {
            return "Unknown (" + installerPackageName + ")";
        } else {
            return "Unknown";
        }
    }

    public static String getEnabledText(ApplicationInfo applicationInfo) {
        return applicationInfo.enabled ? "Enabled" : "Disabled";
    }

    public static Date getFirstInstalled(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return new Date(getPackageInfo(packageManager, applicationInfo).firstInstallTime);
    }

    public static String getFirstInstalledText(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return getDateFormat().format(getFirstInstalled(packageManager, applicationInfo));
    }

    public static Date getLastUpdated(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return new Date(getPackageInfo(packageManager, applicationInfo).lastUpdateTime);
    }

    public static String getLastUpdatedText(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return getDateFormat().format(getLastUpdated(packageManager, applicationInfo));
    }

    public static Date getLastUsed(UsageStatsManager usageStatsManager, ApplicationInfo applicationInfo, boolean reload) {
        Map<String, Long> lastUsedEpochs = getLastUsedEpochs(usageStatsManager, reload);
        Long epoch = lastUsedEpochs.getOrDefault(applicationInfo.packageName, 0L);
        return new Date(epoch == null ? 0L : epoch);
    }

    public static String getLastUsedText(UsageStatsManager usageStatsManager, ApplicationInfo applicationInfo, boolean reload) {
        Date lastUsed = getLastUsed(usageStatsManager, applicationInfo, reload);
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.YEAR, -2);
        if (lastUsed.before(calendar.getTime())) {
            return "Unknown";
        }
        return getDateFormat().format(lastUsed);
    }

    private static Map<String, Long> getLastUsedEpochs(UsageStatsManager usageStatsManager, boolean reload) {
        if (lastUsedEpochsCache == null || reload) {
            Calendar calendar = Calendar.getInstance();
            long endTime = calendar.getTimeInMillis();
            calendar.add(Calendar.YEAR, -2);
            long startTime = calendar.getTimeInMillis();
            lastUsedEpochsCache = usageStatsManager.queryAndAggregateUsageStats(startTime, endTime).entrySet().stream()
                    .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().getLastTimeUsed()));
        }
        return lastUsedEpochsCache;
    }

    public static String getVersionText(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return getPackageInfo(packageManager, applicationInfo).versionName;
    }

    public static List<String> getPermissions(PackageManager packageManager, ApplicationInfo applicationInfo, boolean grantedPermissionsOnly) throws PackageManager.NameNotFoundException {
        List<String> permissions = new ArrayList<>();
        PackageInfo packageInfo = getPackageInfo(packageManager, applicationInfo);
        String[] requestedPermissions = packageInfo.requestedPermissions;
        if (requestedPermissions == null) {
            return permissions;
        }
        for (int i = 0; i < requestedPermissions.length; i++) {
            String requestedPermission = requestedPermissions[i];
            if (!requestedPermission.startsWith("android.permission")) {
                continue;
            }
            boolean granted = (packageInfo.requestedPermissionsFlags[i] & PackageInfo.REQUESTED_PERMISSION_GRANTED) != 0;
            if (grantedPermissionsOnly && !granted) {
                continue;
            }
            permissions.add(requestedPermission);
        }
        return permissions;
    }

    public static long getApkSize(ApplicationInfo applicationInfo) {
        return new File(applicationInfo.publicSourceDir).length();
    }

    public static String getApkSizeText(Context context, ApplicationInfo applicationInfo) {
        return Formatter.formatShortFileSize(context, getApkSize(applicationInfo));
    }

    public static StorageUsage getStorageUsage(Context context, ApplicationInfo applicationInfo) {
        StorageUsage storageUsage = new StorageUsage();
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            Log.w(TAG, "Unable to calculate storage usage (requires API " + Build.VERSION_CODES.O + ")");
            return storageUsage;
        }
        final StorageManager storageManager = (StorageManager) context.getSystemService(Context.STORAGE_SERVICE);
        final StorageStatsManager storageStatsManager = (StorageStatsManager) context.getSystemService(Context.STORAGE_STATS_SERVICE);
        for (StorageVolume storageVolume : storageManager.getStorageVolumes()) {
            if (Environment.MEDIA_MOUNTED.equals(storageVolume.getState())) {
                String uuidStr = storageVolume.getUuid();
                UUID uuid;
                try {
                    uuid = uuidStr == null ? StorageManager.UUID_DEFAULT : UUID.fromString(uuidStr);
                } catch (IllegalArgumentException e) {
                    Log.w(TAG, "Could not parse UUID " + uuidStr + " for calculating storage usage");
                    continue;
                }
                try {
                    StorageStats storageStats = storageStatsManager.queryStatsForPackage(uuid, applicationInfo.packageName, myUserHandle());
                    storageUsage.increaseDataBytes(storageStats.getAppBytes());
                    storageUsage.increaseDataBytes(storageStats.getCacheBytes());
                    storageUsage.increaseDataBytes(storageStats.getDataBytes());
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                        storageUsage.increaseExternalCacheBytes(storageStats.getExternalCacheBytes());
                    }
                } catch (PackageManager.NameNotFoundException | IOException e) {
                    Log.e(TAG, "Unable to process storage usage for " + applicationInfo.packageName + " on " + uuid, e);
                }
            }
        }

        return storageUsage;
    }

    private static PackageInfo getPackageInfo(PackageManager packageManager, ApplicationInfo applicationInfo) throws PackageManager.NameNotFoundException {
        return packageManager.getPackageInfo(applicationInfo.packageName, PackageManager.GET_PERMISSIONS);
    }

    private static DateFormat getDateFormat() {
        return DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT, Locale.getDefault());
    }

    public static class StorageUsage {
        private long appBytes;
        private long cacheBytes;
        private long dataBytes;
        private long externalCacheBytes;
        private long totalBytes;

        public StorageUsage() {}

        public StorageUsage(long appBytes, long cacheBytes, long dataBytes, long externalCacheBytes) {
            this.appBytes = appBytes;
            this.cacheBytes = cacheBytes;
            this.dataBytes = dataBytes;
            this.externalCacheBytes = externalCacheBytes;
            recalculateTotal();
        }

        public long getAppBytes() {
            return appBytes;
        }

        public void setAppBytes(long appBytes) {
            this.appBytes = appBytes;
            recalculateTotal();
        }

        public void incrementAppBytes(long appBytes) {
            this.appBytes += appBytes;
            recalculateTotal();
        }

        public long getCacheBytes() {
            return cacheBytes;
        }

        public void setCacheBytes(long cacheBytes) {
            this.cacheBytes = cacheBytes;
            recalculateTotal();
        }

        public void incrementCacheBytes(long cacheBytes) {
            this.cacheBytes += cacheBytes;
            recalculateTotal();
        }

        public long getDataBytes() {
            return dataBytes;
        }

        public void setDataBytes(long dataBytes) {
            this.dataBytes = dataBytes;
            recalculateTotal();
        }

        public void increaseDataBytes(long dataBytes) {
            this.dataBytes += dataBytes;
            recalculateTotal();
        }

        public long getExternalCacheBytes() {
            return externalCacheBytes;
        }

        public void setExternalCacheBytes(long externalCacheBytes) {
            this.externalCacheBytes = externalCacheBytes;
            recalculateTotal();
        }

        public void increaseExternalCacheBytes(long externalCacheBytes) {
            this.externalCacheBytes += externalCacheBytes;
            recalculateTotal();
        }

        public long getTotalBytes() {
            return totalBytes;
        }

        private void recalculateTotal() {
            totalBytes = appBytes + cacheBytes + dataBytes + externalCacheBytes;
        }
    }
}

File: app/src/main/kotlin/com/github/keeganwitt/applist/ApplicationInfoUtils.kt
package com.github.keeganwitt.applist

import android.app.usage.StorageStatsManager
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import android.os.Environment
import android.os.Process
import android.os.storage.StorageManager
import android.text.format.Formatter
import android.util.Log
import com.google.firebase.crashlytics.FirebaseCrashlytics
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.UUID

object ApplicationInfoUtils {
    private val TAG = ApplicationInfoUtils::class.java.simpleName
    private val httpClient = OkHttpClient()
    private val existsInAppStoreCache: MutableMap<String, Boolean> = HashMap()
    private var lastUsedEpochsCache: Map<String, Long>? = null
    private val dateFormat = SimpleDateFormat.getDateTimeInstance()

    @JvmStatic
    fun getPackageInstaller(
        packageManager: PackageManager,
        applicationInfo: ApplicationInfo
    ): String? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            try {
                packageManager.getInstallSourceInfo(applicationInfo.packageName).installingPackageName
            } catch (e: PackageManager.NameNotFoundException) {
                val message = "Could not get package installer"
                Log.e(TAG, message + " for " + applicationInfo.packageName, e)
                FirebaseCrashlytics.getInstance().log(message)
                FirebaseCrashlytics.getInstance().recordException(e)
                null
            }
        } else {
            @Suppress("DEPRECATION")
            packageManager.getInstallerPackageName(applicationInfo.packageName)
        }
    }

    @JvmStatic
    fun getPackageInstallerName(installerPackageName: String?): String {
        return when (installerPackageName) {
            "com.amazon.venezia" -> "Amazon Appstore"
            "com.google.android.packageinstaller" -> "APK"
            "cm.aptoide.pt" -> "Aptoide"
            "org.fdroid.fdroid" -> "F-Droid"
            "net.rim.bb.appworld" -> "Blackberry World"
            "com.farsitel.bazaar" -> "Cafe Bazaar"
            "com.sec.android.app.samsungapps" -> "Galaxy Store"
            "com.android.vending" -> "Google Play"
            "com.huawei.appmarket" -> "Huawei App Gallery"
            "com.xiaomi.market" -> "Mi Store"
            "com.oneplus.backuprestore" -> "OnePlus Clone Phone"
            "com.sec.android.easyMover" -> "Samsung Smart Switch"
            "com.slideme.sam.manager" -> "SlideME Marketplace"
            "com.tencent.android.qqdownloader" -> "Tencent Appstore"
            "com.yandex.store" -> "Yandex Appstore"
            null -> "Unknown"
            else -> "Unknown ($installerPackageName)"
        }
    }

    @JvmStatic
    fun getEnabledText(context: Context, applicationInfo: ApplicationInfo): String {
        return if (applicationInfo.enabled) context.getString(R.string.enabled) else context.getString(R.string.disabled)
    }

    @JvmStatic
    fun getApkSizeText(context: Context, applicationInfo: ApplicationInfo): String {
        return Formatter.formatShortFileSize(context, getApkSize(applicationInfo))
    }

    @JvmStatic
    fun isAppArchived(applicationInfo: ApplicationInfo): Boolean? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM) {
            applicationInfo.isArchived
        } else {
            null
        }
    }

    @JvmStatic
    fun getAppIsArchivedText(context: Context, applicationInfo: ApplicationInfo): String {
        val archived = isAppArchived(applicationInfo)
        return when (archived) {
            null -> context.getString(R.string.unknown)
            true -> context.getString(R.string.archived)
            false -> context.getString(R.string.installed)
        }
    }

    @JvmStatic
    fun getExistsInAppStoreText(context: Context, packageManager: PackageManager, applicationInfo: ApplicationInfo): String {
        val exists = existsInAppStore(packageManager, applicationInfo)
        return when (exists) {
            null -> context.getString(R.string.unknown)
            true -> {
                val link = getAppStoreLink(applicationInfo)
                "<a href=\"$link\">${context.getString(R.string.boolean_true)}</a>"
            }
            false -> context.getString(R.string.boolean_false)
        }
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getFirstInstalled(packageManager: PackageManager, applicationInfo: ApplicationInfo): Date {
        return Date(getPackageInfo(packageManager, applicationInfo).firstInstallTime)
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getFirstInstalledText(packageManager: PackageManager, applicationInfo: ApplicationInfo): String {
        return dateFormat.format(getFirstInstalled(packageManager, applicationInfo))
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getLastUpdated(packageManager: PackageManager, applicationInfo: ApplicationInfo): Date {
        return Date(getPackageInfo(packageManager, applicationInfo).lastUpdateTime)
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getLastUpdatedText(packageManager: PackageManager, applicationInfo: ApplicationInfo): String {
        return dateFormat.format(getLastUpdated(packageManager, applicationInfo))
    }

    @JvmStatic
    fun getLastUsed(usageStatsManager: UsageStatsManager, applicationInfo: ApplicationInfo, reload: Boolean): Date {
        val lastUsedEpochs = getLastUsedEpochs(usageStatsManager, reload)
        val epoch = lastUsedEpochs[applicationInfo.packageName] ?: 0L
        return Date(epoch)
    }

    fun getLastUsedText(usageStatsManager: UsageStatsManager, applicationInfo: ApplicationInfo, reload: Boolean): String {
        val lastUsed = getLastUsed(usageStatsManager, applicationInfo, reload)
        val calendar = Calendar.getInstance().apply {
            add(Calendar.YEAR, -2)
        }
        return if (lastUsed.before(calendar.time)) {
            "Unknown"
        } else {
            dateFormat.format(lastUsed)
        }
    }

    private fun getLastUsedEpochs(usageStatsManager: UsageStatsManager, reload: Boolean): Map<String, Long> {
        if (lastUsedEpochsCache == null || reload) {
            val calendar = Calendar.getInstance()
            val endTime = calendar.timeInMillis
            calendar.add(Calendar.YEAR, -2)
            val startTime = calendar.timeInMillis
            lastUsedEpochsCache = usageStatsManager.queryAndAggregateUsageStats(startTime, endTime)
                .mapValues { it.value.lastTimeUsed }
        }
        return lastUsedEpochsCache!!
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getVersionText(packageManager: PackageManager, applicationInfo: ApplicationInfo): String? {
        return getPackageInfo(packageManager, applicationInfo).versionName
    }

    @JvmStatic
    @Throws(PackageManager.NameNotFoundException::class)
    fun getPermissions(packageManager: PackageManager, applicationInfo: ApplicationInfo, grantedPermissionsOnly: Boolean): List<String> {
        val packageInfo = getPackageInfo(packageManager, applicationInfo)
        val requestedPermissions = packageInfo.requestedPermissions ?: return emptyList()
        val permissionFlags = packageInfo.requestedPermissionsFlags ?: return emptyList()

        return requestedPermissions.mapIndexedNotNull { i, permission ->
            if (permission.startsWith("android.permission")) {
                val granted = (permissionFlags[i] and PackageInfo.REQUESTED_PERMISSION_GRANTED) != 0
                if (grantedPermissionsOnly && !granted) {
                    null
                } else {
                    permission
                }
            } else {
                null
            }
        }
    }

    @JvmStatic
    fun getApkSize(applicationInfo: ApplicationInfo): Long {
        return File(applicationInfo.publicSourceDir).length()
    }

    @JvmStatic
    fun getStorageUsage(context: Context, applicationInfo: ApplicationInfo): StorageUsage {
        val storageUsage = StorageUsage()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            val message = "Unable to calculate storage usage (requires API " + Build.VERSION_CODES.O + ")"
            Log.w(TAG, message)
            return storageUsage
        }

        val storageManager = context.getSystemService(Context.STORAGE_SERVICE) as StorageManager
        val storageStatsManager = context.getSystemService(Context.STORAGE_STATS_SERVICE) as StorageStatsManager

        storageManager.storageVolumes.forEach { storageVolume ->
            if (Environment.MEDIA_MOUNTED == storageVolume.state) {
                val uuidStr = storageVolume.uuid
                val uuid: UUID = try {
                    if (uuidStr == null) StorageManager.UUID_DEFAULT else UUID.fromString(uuidStr)
                } catch (_: IllegalArgumentException) {
                    val message = "Could not parse UUID $uuidStr for calculating storage usage. This is a known issue on some devices/storage volumes."
                    Log.w(TAG, message)
                    return@forEach
                }

                try {
                    val storageStats = storageStatsManager.queryStatsForPackage(uuid, applicationInfo.packageName, Process.myUserHandle())
                    storageUsage.increaseAppBytes(storageStats.appBytes)
                    storageUsage.increaseCacheBytes(storageStats.cacheBytes)
                    storageUsage.increaseDataBytes(storageStats.dataBytes)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                        storageUsage.increaseExternalCacheBytes(storageStats.externalCacheBytes)
                    }
                } catch (e: SecurityException) {
                    val message = "Missing storage permission"
                    Log.e(TAG, message, e)
                    FirebaseCrashlytics.getInstance().log(message)
                    FirebaseCrashlytics.getInstance().recordException(e)
                } catch (e: Exception) {
                    val message = "Unable to process storage usage"
                    Log.e(TAG, message + " for ${applicationInfo.packageName} on $uuid", e)
                    FirebaseCrashlytics.getInstance().log(message)
                    FirebaseCrashlytics.getInstance().recordException(e)
                }
            }
        }

        return storageUsage
    }

    fun existsInAppStore(packageManager: PackageManager, applicationInfo: ApplicationInfo): Boolean? {
        val id = applicationInfo.packageName
        if (existsInAppStoreCache.containsKey(id)) {
            return existsInAppStoreCache[id]
        }

        if ("Google Play" != getPackageInstallerName(getPackageInstaller(packageManager, applicationInfo))) {
            return null
        }

        val url = getAppStoreLink(applicationInfo)
        val request = Request.Builder().url(url!!).build()
        Log.d(TAG, "Querying $url")

        return try {
            httpClient.newCall(request).execute().use { response ->
                val code = response.code
                val exists = code in 200..299
                if (!exists) {
                    Log.d(TAG, "HTTP response code for $url was $code")
                }
                existsInAppStoreCache[id] = exists
                exists
            }
        } catch (e: IOException) {
            val message = "Unable to make HTTP request to $url"
            Log.e(TAG, message, e)
            FirebaseCrashlytics.getInstance().log(message)
            FirebaseCrashlytics.getInstance().recordException(e)
            null
        }
    }

    fun getAppStoreLink(applicationInfo: ApplicationInfo): String? {
        return "https://play.google.com/store/apps/details?id=${applicationInfo.packageName}"
    }

    @Throws(PackageManager.NameNotFoundException::class)
    private fun getPackageInfo(packageManager: PackageManager, applicationInfo: ApplicationInfo): PackageInfo {
        val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            PackageManager.GET_PERMISSIONS or PackageManager.GET_SIGNING_CERTIFICATES
        } else {
            @Suppress("DEPRECATION")
            PackageManager.GET_PERMISSIONS or PackageManager.GET_SIGNATURES
        }
        return packageManager.getPackageInfo(applicationInfo.packageName, flags)
    }
}
